### PR TITLE
Don't suggest `&mut mut`

### LINKED ISSUE #57431: `&mut &T` coerced to `&T` suggests `&mut mut x`
This code:
```rust
struct X;
impl X {
    fn mutate(&mut self) {}
}

fn main() {
    let term = Some(X);
    let term = match &term {
        Some(term) => &mut term,
        None => term.as_ref().unwrap(),
    };
    term.mutate();
}
```

Produces these errors:
```
error[E0596]: cannot borrow `term` as mutable, as it is not declared as mutable
 --> src/main.rs:9:23
  |
9 |         Some(term) => &mut term,
  |                       ^^^^^^^^^
  |                       |
  |                       cannot borrow as mutable
  |                       try removing `&mut` here

error[E0596]: cannot borrow `*term` as mutable, as it is behind a `&` reference
  --> src/main.rs:12:5
   |
9  |         Some(term) => &mut term,
   |                       --------- help: consider changing this to be a mutable reference: `&mut mut term`
...
12 |     term.mutate();
   |     ^^^^ `term` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error: aborting due to 2 previous errors
```

Note that the second error suggests changing `&mut term` to `&mut mut term`, which obviously won't fix the issue.

### PR DESCRIPTION
**Pull request by saleemjaffer**
fixes #57431

### CODE REVIEW

**Comment by saleemjaffer** [comment]
@spastorino  I have added a `coerced` field in `BindByValue`. Things only compile now. What do you think?

**Comment by spastorino** [comment]
Yeah, continue in this direction. I'm not entirely sure if adding the attribute there is the best idea as we discussed on Zulip. Anyway, keep this going, now you need to assign coerced to true in the right spot and handle that in the diagnostics part :).

**Comment by matthewjasper** [comment]
Looking at the way this message is generated, I think that a larger change is probably deserved here.
If you just want to avoid the incorrect message though. I would suggest moving the flag to `mir::Ravlue::Ref` or `mir::BorrowKind`. You can then retrieve the flag by inspecting `mir.basic_blocks()[locations[0].block].statements.get(locations[0].statement_index)` here:

https://github.com/rust-lang/rust/blob/316a391dcb7d66dc25f1f9a4ec9d368ef7615005/src/librustc_mir/borrow_check/mutability_errors.rs#L572

**Comment by saleemjaffer** [comment]
@matthewjasper This is the the output of `mir.basic_blocks()[locations[0].block].statements.get(locations[0].statement_index)`: 
`Some( _3 = &(*_5),)`.

I have no idea how to make sense of what this means. Any help?

**Comment by matthewjasper** [comment]
The `Debug` implementation for `Statement` pretty prints the statement. That's the output for something like

```text
Statement {
    source_info: / * ... */,
    kind: StatementKind::Assign(
        _3,
        Rvalue::Ref('_, BorrowKind::Shared, *_5),
    ),
}
```

**Comment by saleemjaffer** [comment]
@matthewjasper I tried:
`println!("{:?}", mir.basic_blocks()[locations[0].block].statements.get(locations[0].statement_index));`

`dbg!("{:?}", mir.basic_blocks()[locations[0].block].statements.get(locations[0].statement_index));`

The above 2 print `Some(_3 = &(*_5))`.

I tried:
`debug!("{:?}", mir.basic_blocks()[locations[0].block].statements.get(locations[0].statement_index));`

This does not print anything. Am I missing out on something?

**Comment by saleemjaffer** [comment]
@matthewjasper 

I tried: `println!("{:#?}", mir.basic_blocks()[locations[0].block].statements.get(locations[0].statement_index));`

Should I be doing something else?

**Comment by matthewjasper** [comment]
There's no way to structurally pretty print. You can `match` on the some case and print out the `.kind`, which has a derived Debug implementation (but it's fields all have hand written debug implementations). You can have a look at `https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/enum.StatementKind.html` to see how this type looks.

**Comment by matthewjasper** [comment]
Alternatively you could look in `mir/mod.rs` and replace the manual debug implemenations with the derived ones while you working (and revert later). 
https://github.com/rust-lang/rust/blob/d4a32d504a5aa49b951bfc70602a9615cb772acf/src/librustc/mir/mod.rs#L1868
https://github.com/rust-lang/rust/blob/d4a32d504a5aa49b951bfc70602a9615cb772acf/src/librustc/mir/mod.rs#L2159
https://github.com/rust-lang/rust/blob/d4a32d504a5aa49b951bfc70602a9615cb772acf/src/librustc/mir/mod.rs#L2302
https://github.com/rust-lang/rust/blob/d4a32d504a5aa49b951bfc70602a9615cb772acf/src/librustc/mir/mod.rs#L2475

File: src/librustc/middle/expr_use_visitor.rs [hunk 1]
```diff
@@ -828,7 +828,7 @@ impl<'a, 'gcx, 'tcx> ExprUseVisitor<'a, 'gcx, 'tcx> {
                 match bm {
                     ty::BindByReference(..) =>
                         mode.lub(BorrowingMatch),
-                    ty::BindByValue(..) => {
+                    ty::BindByValue{..} => {
```
Comment (varkor) [inline src/librustc/middle/expr_use_visitor.rs]:
```suggestion
                    ty::BindByValue { .. } => {
```
(here and elsewhere)

**Comment by jonas-schievink** [comment]
Ping from triage @saleemjaffer, there's merge conflicts waiting to be resolved

**Comment by pnkfelix** [comment]
I'm going to close this PR since it seems to be inactive and incomplete. (If nothing else, it lacks a regression test.)

as @matthewjasper said, we probably should try to resolve the problem of #57431 via a larger change.

**Pull request closed without merging**